asphr/assert.hpp

```diff
@@ -0,0 +1,130 @@
+//
+// Recommendation: Modern CPUs dynamically predict branch execution paths,
+// typically with accuracy greater than 97%. As a result, annotating every
+// branch in a codebase is likely counterproductive; however, annotating
+// specific branches that are both hot and consistently mispredicted is likely
+// to yield performance improvements.
+#if ASPHR_HAVE_BUILTIN(__builtin_expect) || \
+    (defined(__GNUC__) && !defined(__clang__))
+#define ASPHR_PREDICT_FALSE(x) (__builtin_expect(false || (x), false))
+#define ASPHR_PREDICT_TRUE(x) (__builtin_expect(false || (x), true))
+#else
+#define ASPHR_PREDICT_FALSE(x) (x)
+#define ASPHR_PREDICT_TRUE(x) (x)
+#endif
+
+// the following is inspired by BOOST. license:
+// Boost Software License - Version 1.0 - August 17th, 2003
+
+// Permission is hereby granted, free of charge, to any person or
+// organization obtaining a copy of the software and accompanying
+// documentation covered by this license (the "Software") to use, reproduce,
+// display, distribute, execute, and transmit the Software, and to prepare
+// derivative works of the Software, and to permit third-parties to whom the
+// Software is furnished to do so, all subject to the following:
+
+// The copyright notices in the Software and this entire statement,
+// including the above license grant, this restriction and the following
+// disclaimer, must be included in all copies of the Software, in whole or
+// in part, and all derivative works of the Software, unless such copies or
+// derivative works are solely in the form of machine-executable object code
+// generated by a source language processor.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, TITLE AND
+// NON-INFRINGEMENT. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR ANYONE
+// DISTRIBUTING THE SOFTWARE BE LIABLE FOR ANY DAMAGES OR OTHER LIABILITY,
+// WHETHER IN CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#define PRINT_CERR(x, x_val) std::cerr << x << " = " << x_val << std::endl
+
+// asserts
+#ifndef NDEBUG
+#define ASPHR_ASSERT_MSG(expr, msg)                                         \
+  {                                                                         \
+    if (ASPHR_PREDICT_TRUE(expr)) {                                         \
+      static_cast<void>(0);                                                 \
+    } else {                                                                \
+      std::cerr << "Assertion failed: Expression '" << #expr                \
+                << "' is false in function '" << __func__ << "' location '" \
+                << __FILE__ << ":" << __LINE__ << "': '" << msg << "'.\n";  \
+      std::cerr << boost::stacktrace::stacktrace() << std::endl;            \
```

Comment:
I liked having a function that does this. Not sure why you removed it but I guess you tested it.